horovod/common/operations.cc

```diff
@@ -995,68 +1001,135 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
-          RECORD_EVENT(entries, event_queue, NCCL_REDUCE, stream)
+          RECORD_EVENT(entries, event_queue, NCCL_REDUCESCATTER, stream)
         }
 
-        if (horovod_global.local_rank == 0) {
-          // cudaHostAlloc is significantly slower than malloc.  Pre-allocating
-          // a buffer is not safe since the tensor can be arbitrarily large.
-          host_buffer = malloc(buffer_len);
+        if (num_elements_remaining > 0) {
 
-          CUDA_CHECK(entries, "cudaMemcpyAsync",
-                     cudaMemcpyAsync(host_buffer, buffer_data, buffer_len,
-                                     cudaMemcpyDeviceToHost, stream))
-          // This event must be recorded for the subsequent synchronize.
-          RECORD_EVENT(entries, event_queue, MEMCPY_IN_HOST_BUFFER, stream)
+          // Reduce the remaining data at local_size-1 to append to 
+          // existing buffer
+          NCCL_CHECK(
+              entries, "ncclReduce",
+              ncclReduce(fused_input_data_remainder, buffer_data_remainder,
+                         (size_t)num_elements_remaining,
+                         GetNCCLDataType(first_entry.tensor), ncclSum,
+                         horovod_global.local_size - 1, nccl_comm, stream))
+          if (timeline.Initialized()) {
+            RECORD_EVENT(entries, event_queue, NCCL_REDUCE, stream)
+          }
+        }
 
-          // Synchronize.
-          WAIT_FOR_EVENTS(entries, timeline, event_queue)
+        // cudaHostAlloc is significantly slower than malloc.  Pre-allocating
+        // a buffer is not safe since the tensor can be arbitrarily large.
+        host_buffer = malloc(total_buffer_len);
 
-          ACTIVITY_START_ALL(entries, timeline, MPI_ALLREDUCE)
-          MPI_CHECK(entries, "MPI_Allreduce",
-                    MPI_Allreduce(MPI_IN_PLACE, host_buffer, (int)num_elements,
-                                  GetMPIDataType(first_entry.tensor), MPI_SUM,
-                                  horovod_global.cross_comm))
-          ACTIVITY_END_ALL(entries, timeline)
+        CUDA_CHECK(entries, "cudaMemcpyAsync",
+                   cudaMemcpyAsync(host_buffer, buffer_data_at_rank_offset,
+                                   total_buffer_len, cudaMemcpyDeviceToHost,
+                                   stream))
+        // This event must be recorded for the subsequent synchronize.
+        RECORD_EVENT(entries, event_queue, MEMCPY_IN_HOST_BUFFER, stream)
 
-          CUDA_CHECK(entries, "cudaMemcpyAsync",
-                     cudaMemcpyAsync(buffer_data, host_buffer, buffer_len,
-                                     cudaMemcpyHostToDevice, stream))
-          if (timeline.Initialized()) {
-            RECORD_EVENT(entries, event_queue, MEMCPY_OUT_HOST_BUFFER, stream)
-          }
+        // Synchronize.
+        WAIT_FOR_EVENTS(entries, timeline, event_queue)
+
+        ACTIVITY_START_ALL(entries, timeline, MPI_ALLREDUCE)
+        MPI_CHECK(entries, "MPI_Allreduce",
```

Comment:
not support fp16, so should add self define fp16 add func for mpi

Reply:
BUG here. 
last rank will MPI_Allreduce with total_num_elements, while other ranks with total_num_elements-1.
can't combine them to one call.

Reply:
how to local test this  hierarchical?